l5kit/l5kit/simulation/unroll.py

```diff
@@ -0,0 +1,145 @@
+    :param output_dict: the output of the agent model
+    :return:
+    """
+
+    agents_update_dict: Dict[Tuple[int, int]: np.ndarray] = {}
+    input_dict = {k: v.cpu().numpy() for k, v in input_dict.items()}
+    output_dict = {k: v.cpu().numpy() for k, v in output_dict.items()}
+
+    world_from_agent = input_dict["world_from_agent"]
+    yaw = input_dict["yaw"]
+    pred_trs = transform_points(output_dict["positions"][:, :1], world_from_agent)[:, 0]
+    pred_yaws = yaw + output_dict["yaws"][:, 0]
+
+    next_agents = np.zeros(len(yaw), dtype=AGENT_DTYPE)
+    next_agents["centroid"] = pred_trs
+    next_agents["yaw"] = pred_yaws
+    next_agents["track_id"] = input_dict["track_id"]
+    next_agents["extent"] = input_dict["extent"]
+    next_agents["label_probabilities"][:, 3] = 1
+
+    for scene_idx, next_agent in zip(input_dict["scene_index"], next_agents):
+        agents_update_dict[(scene_idx, next_agent["track_id"])] = np.expand_dims(next_agent, 0)
+    dataset.set_agents(frame_idx, agents_update_dict)
+
+
+def update_ego(dataset: SimulationDataset, frame_idx: int, input_dict: Dict[str, torch.Tensor],
+                  output_dict: Dict[str, torch.Tensor]) -> None:
+    """Update ego across scenes for the given frame index.
+
+    :param dataset: The simulation dataset
+    :param frame_idx: index of the frame to modify
+    :param input_dict: the input to the ego model
+    :param output_dict: the output of the ego model
+    :return:
+    """
+    input_dict = {k: v.cpu().numpy() for k, v in input_dict.items()}
+    output_dict = {k: v.cpu().numpy() for k, v in output_dict.items()}
+
+    world_from_agent = input_dict["world_from_agent"]
+    yaw = input_dict["yaw"]
+    pred_trs = transform_points(output_dict["positions"][:, :1], world_from_agent)
+    pred_yaws = yaw + output_dict["yaws"][:, :1]
+
+    dataset.set_ego_for_frame(frame_idx, 0, pred_trs, pred_yaws)
+
+
+@torch.no_grad()
+def unroll(
+    cfg: Dict,
+    model_ego: torch.nn.Module,
+    model_agents: torch.nn.Module,
+    dataset: SimulationDataset,
+    start_frame_index: int = 0,
+    end_frame_index: Optional[int] = None,
+    ) -> None:
+
+    if end_frame_index is None:
+        end_frame_index = len(dataset)
+
+    for frame_index in tqdm(range(start_frame_index, end_frame_index)):
+        next_frame_index = frame_index + 1
+
+        # AGENTS
+        agents_input = dataset.rasterise_agents_frame_batch(frame_index)
+        agents_input_dict = default_collate(list(agents_input.values()))
+        agents_output_dict = model_agents(agents_input_dict)
+
+        # EGO
+        ego_input = dataset.rasterise_frame_batch(frame_index)
+        ego_input_dict = default_collate(ego_input)
+        ego_output_dict = model_ego(ego_input_dict)
+
+        if next_frame_index != end_frame_index:
+            update_agents(dataset, next_frame_index, agents_input_dict, agents_output_dict)
+            update_ego(dataset, next_frame_index, ego_input_dict, ego_output_dict)
```

Comment:
shall we create a class `ClosedLoopSimulator` as in jadoo? https://github.com/lyft/avjadoo/blob/00348d969ddfdfc153d5cb0279bf6ff8b3688696/avjadoo/data/ddp_scene/sim/closed_loop_simulation.py#L213

Reply:
I'm not sure I'll buy us anything, but let me try